src/services/formatting/rules.ts

```diff
@@ -649,6 +662,28 @@ module ts.formatting {
             return context.TokensAreOnSameLine();
         }
 
+        static IsEndOfDecoratorContextOnSameLine(context: FormattingContext): boolean {
+            return context.TokensAreOnSameLine() &&
+                context.contextNode.decorators &&
+                Rules.NodeIsInDecoratorContext(context.currentTokenParent) &&
+                !Rules.NodeIsInDecoratorContext(context.nextTokenParent);
+        }
+
+        static NodeIsInDecoratorContext(node: Node): boolean {
+            if (node.parserContextFlags & ParserContextFlags.Decorator) {
+                return true;
+            }
+            while (node) {
+                if (isExpression(node)) {

```

Comment:
Based on our offline discussion, I changed the logic in `isExpression` in #2659. This should no longer be an issue.